stetho/src/main/java/com/facebook/stetho/inspector/network/ResponseBodyFileManager.java

```diff
@@ -83,4 +125,49 @@ public OutputStream openResponseBodyFile(String requestId, boolean base64Encode)
   private static String getFilename(String requestId) {
     return FILENAME_PREFIX + requestId;
   }
+
+  /**
+   * Associates an asynchronous pretty printer with a response request id
+   * The pretty printer will be used to pretty print the response body that has
+   * the particular request id
+   *
+   * @param requestId Unique identifier for the response
+   * as per {@link NetworkEventReporter.InspectorResponse#requestId()}
+   * @param asyncPrettyPrinter Asynchronous Pretty Printer to pretty print the response body
+   */
+  public void associateAsyncPrettyPrinterWithId(
+      String requestId,
+      AsyncPrettyPrinter asyncPrettyPrinter) {
+    if (mRequestIdMap.containsKey(requestId)) {
+      throw new IllegalArgumentException("cannot associate different " +
+          "pretty printers with the same request id: "+requestId);
+    }
+    mRequestIdMap.put(requestId, asyncPrettyPrinter);
+  }
+
+  private class AsyncPrettyPrintingCallable implements Callable<String> {
+    private final InputStream mInputStream;
+    private final AsyncPrettyPrinter mAsyncPrettyPrinter;
+
+    public AsyncPrettyPrintingCallable(
+        InputStream in,
+        AsyncPrettyPrinter asyncPrettyPrinter) {
+      mInputStream = in;
+      mAsyncPrettyPrinter = asyncPrettyPrinter;
+    }
+
+    @Override
+    public String call() throws IOException {
+      return prettyPrintContent(mInputStream, mAsyncPrettyPrinter);
+    }
+
+    private String prettyPrintContent(InputStream in, AsyncPrettyPrinter asyncPrettyPrinter)
+        throws IOException {
+      ByteArrayOutputStream out = new ByteArrayOutputStream();
+      PrintWriter writer = new PrintWriter(out);
+      asyncPrettyPrinter.printTo(writer, in);
```

Comment:
This needs to have timeout, and should use thread interrupt to attempt to unblock the caller.  A timeout can be implemented using an Android `Handler` and `sendMessageDelayed`.  It's safe to use the main thread for this since this is a debug tool:

```java
  private final Handler mPrettyPrinterTimeoutHandler = new Handler(Looper.mainLooper());

  ...

  final Thread myThread = Thread.currentThread();
  mHandler.postRunnableDelayed(() -> myThread.interrupt(), SOME_DELAY);
```